File: agent/packages/schema/src/spirit-eye.ts
import { Type, type Static } from "@sinclair/typebox";

import { validate, type ValidationResult } from "./validate.js";

export const SpiritEyePositionV1 = Type.Object(
  {
    x: Type.Number(),
    y: Type.Number(),
    z: Type.Number(),
  },
  { additionalProperties: false },
);
export type SpiritEyePositionV1 = Static<typeof SpiritEyePositionV1>;

export const SpiritEyeMigrateReasonV1 = Type.Union([
  Type.Literal("usage_pressure"),
  Type.Literal("periodic_drift"),
]);
export type SpiritEyeMigrateReasonV1 = Static<typeof SpiritEyeMigrateReasonV1>;

export const SpiritEyeMigrateV1 = Type.Object(
  {
    v: Type.Literal(1),
    eye_id: Type.String({ minLength: 1, maxLength: 160 }),
    from: SpiritEyePositionV1,
    to: SpiritEyePositionV1,
    reason: SpiritEyeMigrateReasonV1,
    usage_pressure: Type.Number({ minimum: 0 }),
    tick: Type.Integer({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type SpiritEyeMigrateV1 = Static<typeof SpiritEyeMigrateV1>;

export const SpiritEyeDiscoveredV1 = Type.Object(
  {
    v: Type.Literal(1),
    eye_id: Type.String({ minLength: 1, maxLength: 160 }),
    character_id: Type.String({ minLength: 1, maxLength: 160 }),
    pos: SpiritEyePositionV1,
    zone: Type.Optional(Type.String({ minLength: 1, maxLength: 160 })),
    qi_concentration: Type.Number({ minimum: 0 }),
    discovered_at_tick: Type.Integer({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type SpiritEyeDiscoveredV1 = Static<typeof SpiritEyeDiscoveredV1>;

export const SpiritEyeUsedForBreakthroughV1 = Type.Object(
  {
    v: Type.Literal(1),
    eye_id: Type.String({ minLength: 1, maxLength: 160 }),
    character_id: Type.String({ minLength: 1, maxLength: 160 }),
    realm_from: Type.String({ minLength: 1, maxLength: 64 }),
    realm_to: Type.String({ minLength: 1, maxLength: 64 }),
    usage_pressure: Type.Number({ minimum: 0 }),
    tick: Type.Integer({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type SpiritEyeUsedForBreakthroughV1 = Static<typeof SpiritEyeUsedForBreakthroughV1>;

export const SpiritEyeCoordinateNoteV1 = Type.Object(
  {
    v: Type.Literal(1),
    eye_id: Type.String({ minLength: 1, maxLength: 160 }),
    owner_character_id: Type.String({ minLength: 1, maxLength: 160 }),
    pos: SpiritEyePositionV1,
    zone: Type.Optional(Type.String({ minLength: 1, maxLength: 160 })),
    qi_concentration: Type.Number({ minimum: 0 }),
    discovered_at_tick: Type.Integer({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type SpiritEyeCoordinateNoteV1 = Static<typeof SpiritEyeCoordinateNoteV1>;

export const DeathInsightSpiritEyeV1 = Type.Object(
  {
    eye_id: Type.String({ minLength: 1, maxLength: 160 }),
    zone: Type.Optional(Type.String({ minLength: 1, maxLength: 160 })),
    pos: SpiritEyePositionV1,
    qi_concentration: Type.Number({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type DeathInsightSpiritEyeV1 = Static<typeof DeathInsightSpiritEyeV1>;

export function validateSpiritEyeMigrateV1Contract(data: unknown): ValidationResult {
  return validate(SpiritEyeMigrateV1, data);
}

export function validateSpiritEyeDiscoveredV1Contract(data: unknown): ValidationResult {
  return validate(SpiritEyeDiscoveredV1, data);
}

export function validateSpiritEyeUsedForBreakthroughV1Contract(
  data: unknown,
): ValidationResult {
  return validate(SpiritEyeUsedForBreakthroughV1, data);
}

File: server/src/schema/spirit_eye.rs
use serde::{Deserialize, Serialize};

use super::death_insight::DeathInsightPositionV1;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct SpiritEyePositionV1 {
    pub x: f64,
    pub y: f64,
    pub z: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SpiritEyeMigrateReasonV1 {
    UsagePressure,
    PeriodicDrift,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct SpiritEyeMigrateV1 {
    pub v: u8,
    pub eye_id: String,
    pub from: SpiritEyePositionV1,
    pub to: SpiritEyePositionV1,
    pub reason: SpiritEyeMigrateReasonV1,
    pub usage_pressure: f64,
    pub tick: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct SpiritEyeDiscoveredV1 {
    pub v: u8,
    pub eye_id: String,
    pub character_id: String,
    pub pos: SpiritEyePositionV1,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub zone: Option<String>,
    pub qi_concentration: f64,
    pub discovered_at_tick: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct SpiritEyeUsedForBreakthroughV1 {
    pub v: u8,
    pub eye_id: String,
    pub character_id: String,
    pub realm_from: String,
    pub realm_to: String,
    pub usage_pressure: f64,
    pub tick: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct SpiritEyeCoordinateNoteV1 {
    pub v: u8,
    pub eye_id: String,
    pub owner_character_id: String,
    pub pos: SpiritEyePositionV1,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub zone: Option<String>,
    pub qi_concentration: f64,
    pub discovered_at_tick: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct DeathInsightSpiritEyeV1 {
    pub eye_id: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub zone: Option<String>,
    pub pos: DeathInsightPositionV1,
    pub qi_concentration: f64,
}

impl From<[f64; 3]> for SpiritEyePositionV1 {
    fn from(pos: [f64; 3]) -> Self {
        Self {
            x: pos[0],
            y: pos[1],
            z: pos[2],
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn spirit_eye_migrate_contract_serializes_snake_case_reason() {
        let payload = SpiritEyeMigrateV1 {
            v: 1,
            eye_id: "spirit_eye:spawn:0".to_string(),
            from: [0.0, 66.0, 0.0].into(),
            to: [640.0, 66.0, 0.0].into(),
            reason: SpiritEyeMigrateReasonV1::UsagePressure,
            usage_pressure: 1.0,
            tick: 72,
        };

        let json = serde_json::to_value(payload).expect("serialize");

        assert_eq!(json["reason"], "usage_pressure");
        assert_eq!(json["eye_id"], "spirit_eye:spawn:0");
    }
}
